### PR TITLE
Split core functionality from CLI

### PR DESCRIPTION
**Pull request by kachkaev**
👋 

I recently needed to use `run-elm` via `require()` rather than by spawning a sub-process and came up with a refactoring that involved splitting core functionality from CLI. At the moment this change is included into a fork called [@kachkaev/run-elm](https://www.npmjs.com/package/@kachkaev/run-elm), but I hope to deprecate that package once we update `run-elm`.

~~~I have marked this PR as work-in-progress because it includes commits from https://github.com/jfairbank/run-elm/pull/13, has some fork-specific commits and also lacks usage instructions for `require()` in README. The tests have been done though, all works!~~~

One thing that bugs me is that I'm using `compileSync` instead of `compile` (they both replace `compileToString` and save one cycle of file write/read). The reason is in https://github.com/rtfeldman/node-elm-compiler/issues/68 (async version calls `process.exit()` on error, so it can't be used when we `require('run-elm')`).

WDYT @jfairbank? ~~~I'm happy to rebase and refactor after merging #13 and getting your general approval.~~~

### CODE REVIEW

**Comment by jfairbank** [comment]
This seems like a good change to make. I'll wait for you to rebase before I review. Thanks!

**Comment by kachkaev** [comment]
Despite that all tests pass, I'm still not sure this PR is worth merging as is. The issue that I see is that we're calling `node-elm-compiler` synchronously, because otherwise we'll get early exit due to https://github.com/rtfeldman/node-elm-compiler/issues/68. Therefore, the main Node.js thread will get blocked and performance will suffer (this only applies when Node API is used). There is no regression in CLI use, just a non-optimality in the new execution mode.

What are your thoughts on this @jfairbank? How is the PR in general?

**Comment by kachkaev** [comment]
I also noticed today that Elm's `Debug.log` is being incorrectly reported in the API mode. Will add a fix to this soon, please don't merge.

**Comment by kachkaev** [comment]
I also managed to fix a few things so that they work on Windows:
https://ci.appveyor.com/project/kachkaev/run-elm/build/27

Things to note:
* stderr / stdout never contain `\r` (even when it's Windows) to make sure there are no unwanted parsing issues in the downstream projects. If someone asks for a nice-looking output for `cmd`, we can look into adding a CLI option for that.

* I had to rename `Aux.elm` into `Helper.elm` in one of the tests, because you can't use `AUX` for file names on Windows 😱  This was giving a [pretty strange exception](https://ci.appveyor.com/project/kachkaev/run-elm/build/22/job/t58m6fbf86aiwag6#L83).

@jfairbank feel free to integrate AppVeyor to the repo just as I did in my fork.

One last thing I'd like to look into is to make a call to elm compiler async despite https://github.com/rtfeldman/node-elm-compiler/issues/68.

File: package.json [hunk 1]
```diff
@@ -6,18 +6,21 @@
...
   "engines": {
     "node": ">=6.0.0"
   },
   "bin": {
-    "run-elm": "lib/index.js"
+    "run-elm": "lib/cli.js"
   },
   "scripts": {
-    "test:clean": "cd test/integration && git clean -d -f -X",
+    "clean:test": "git clean -d -f -X test/integration",
```
Comment (kachkaev) [inline package.json:17]: && are not allowed in npm scripts on Windows; also renamed `test:clean` into `clean:test` for a better clarity

File: package.json [hunk 2]
```diff
@@ -6,18 +6,21 @@
...
     "lint": "eslint src test --ignore-pattern elm-stuff",
+    "prebuild": "rimraf lib",
+    "build": "babel src --out-dir lib",
+    "postbuild": "cpx \"src/*.elm.template\" lib",
```
Comment (kachkaev) [inline package.json:21]: && are not allowed in npm scripts on Windows, so I had to split command into three

File: src/index.js [hunk 3]
```diff
@@ -70,10 +37,10 @@ import sh from 'shelljs';
 
     // ensure --output-name is specified adequately
     if (!outputName.match(/^[a-z_]\w*$/)) {
-      throw new Error(`Provided --output-name \`${outputName}\` is not a valid constant or function name in elm.`);
+      throw new Error(`Provided output name \`${outputName}\` is not a valid constant or function name in elm.`);
```
Comment (kachkaev) [inline src/index.js:42]: not using `--output-name` to avoid confusion in non-CLI usage. Same applies to other args

File: src/index.js [hunk 4]
```diff
@@ -137,52 +104,68 @@ import sh from 'shelljs';
...
+    err.message = message && process.platform === 'win32'
+      ? message.replace(/\r\n/g, '\n')
+      : message;
+    throw err;
   } finally {
     // cleanup
+    await Promise.all([
+      close(elmCompileStdoutFd).catch(() => {}),
+      close(elmCompileStderrFd).catch(() => {}),
+    ]);
```
Comment (kachkaev) [inline src/index.js]: Had to do this to support Windows, otherwise the files were left unclosed and could not be deleted

**Comment by kachkaev** [comment]
The PR should be ready for your review @jfairbank 🎉  Calls to `elm-compile` are async now and both Travis and [AppVeyor#28](https://ci.appveyor.com/project/kachkaev/run-elm/build/28) (Windows) pass!

**Comment by kachkaev** [comment]
ping @jfairbank 🏓 🙂

**Comment by jfairbank** [comment]
Hey, @kachkaev. I've been at a conference this week and am traveling the rest of the week, so I won't be able to check this PR more thoroughly until next week. Sorry for the delay, but I didn't want to keep you in the dark at least.

**Comment by kachkaev** [comment]
@jfairbank when about could you have a look at this? No rush, just curious.

**Comment by jfairbank** [comment]
Sorry again; I've just been super busy with other priorities. I finally looked this over, and it seems reasonable. I'll get a release out later tonight. Thanks!

**Comment by jfairbank** [comment]
v2.3.0 is now out!

**Pull request merged**